### PR TITLE
StreamWindow abstraction + modular window computations

### PR DESCRIPTION
**Pull request by gyfora**
This PR introduces a rework to the current windowing semantics, by major API and Runtime improvements.

The core new abstraction is the StreamWindow which encapsulates the contents of a window in the data stream. Now the WindowedDataStream symbolises transformations on a DataStream of StreamWindows. This allows the users to properly collect, and work with the results of the window transformations.

API changes (for a full overview https://github.com/mbalassi/flink/blob/window/docs/streaming_guide.md#window-operators):

Previously the call `ds.window(...).groupBy(...).reduce(...)` would have returned a stream of reduced values by key in the given window. There was no way of properly obtaining all the (key, value) pairs from the window reduce calls just the flattened windows. This also made it impossible to apply more than one transformations on the same window without having to re-descritize it.

The new api introduces several changes here:

`ds.window(...).groupBy(...).reduceWindow(...)` returns a WindowedDataStream which than can be further transformed by applying transformations on the resulting window. The user can also call `.flatten()` or `.getDiscretizedStream()` to obtain the DataStream of the flattened values or the DataStream of StreamWindows.

The reduce and reduceGroup functions have been renamed to reduceWindow and mapWindow respectively to emphasise the functionality.

Runtime changes:

The windowing runtime has been completely reworked to make the components modular for future improvements:

DataStream -> StreamDiscretizer -> (PreAggregator) -> WindowPartitioner -> Transformation (reduce/map) -> WindowMerge 

This allows easier special-casing for specific window types for maximal performance, such as applying special preaggregators or parallel stream discretisation for global windows (time) as well.

Further tasks after merge:
-Add pre-aggregators for time windows
-Add pre-aggregator for sliding count windows
-Improve test coverage
-Further optimise partitioning-reduce parallelism

### CODE REVIEW

**Comment by senorcarbone** [comment]
+1 for merging asap. It's an improvement on both semantics and performance prospects for windowing.

File: docs/streaming_guide.md [hunk 1]
```diff
@@ -383,9 +389,40 @@ dataStream.groupBy(groupingField).window(Count.of(100)).every(…).max(field);
...
 This will create separate windows for different keys and apply the trigger and eviction policies on a per group basis.
 
+#### Applying multiple transformations on a window
+Using the `WindowedDataStream` abstraction we can apply several transformations one after another on the discretised streams without having to re-discretise it:
+
+~~~java
+dataStream.window(Count.of(1000)).groupBy(firstKey).mapWindow(…).groupBy(secondKey).reduceWindow(…).flatten()
+~~~
+
+The above call would create global windows of 1000 elements group it by the first key and then apply a mapWindow transformation. The resulting windowed stream will then be grouped by the second key and further reduced. The results of the reduce transformation are then flattened.
```
Comment (mbalassi) [inline docs/streaming_guide.md]: Nice example.

File: docs/streaming_guide.md [hunk 2]
```diff
@@ -857,7 +895,7 @@ After installing Docker an image can be pulled for each connector. Containers ca
 For the easiest set up create a jar with all the dependencies of the *flink-streaming-connectors* project.
 
 ~~~batch
-cd /PATH/TO/GIT/flink/flink-staging/flink-streaming-connectors
+cd /PATH/TO/GIT/incubator-flink/flink-addons/flink-streaming-connectors
 mvn assembly:assembly
```
Comment (mbalassi) [inline docs/streaming_guide.md]: An amend here would be nice. :) I'll do it while merging.

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/windowing/StreamWindowTypeInfo.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+	}
+
+	@Override
+	public int getArity() {
+		return innerType.getArity();
+	}
+
+	@Override
+	public Class<StreamWindow<T>> getTypeClass() {
+		// TODO Auto-generated method stub
```
Comment (rmetzger) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/windowing/StreamWindowTypeInfo.java]: hehe

**Pull request closed without merging**